src/twisted/internet/defer.py

```diff
@@ -1400,17 +1425,29 @@ def _inlineCallbacks(result, g, status):
     # loop and the waiting variable solve that by manually unfolding the
     # recursion.
 
-    waiting = [True, # waiting for result?
-               None] # result
+    waiting = [True,  # waiting for result?
+               None]  # result
+
+    if _contextvars:
+        current_context = _contextvars.copy_context()
 
     while 1:
         try:
             # Send the last result back as the result of the yield expression.
             isFailure = isinstance(result, failure.Failure)
-            if isFailure:
-                result = result.throwExceptionIntoGenerator(g)
+
+            if _contextvars:
+                if isFailure:
+                    result = current_context.run(
```

Comment:
Again it seems like making `current_context` into something that duck-types to `Context` would make this implementation shorter and less duplicative.